### PR TITLE
Track which attributes are used during request processing.

### PR DESCRIPTION
**Pull request by geeknoid**
- We track which attributes are referenced for each Mixer Check call and report this to Envoy. Envoy uses this information to properly cache the Check response.

- Fixes a bug in the proto bag code which didn't correctly handle the case where a word was both in the global and the message-level word lists. The message-level word list now takes precedence in all cases.

- Adds the needed support to enable upgrades of the global word list by limiting used of the global word list in generated Attributes messages to what the caller indicated is its maximum support word count.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/985)

### CODE REVIEW

File: pkg/attribute/protoBag.go [hunk 1]
```diff
@@ -50,9 +64,56 @@ func (pb *ProtoBag) Get(name string) (interface{}, bool) {
...
+func (pb *ProtoBag) GetReferencedAttributes() mixerpb.ReferencedAttributes {
+	output := mixerpb.ReferencedAttributes{}
+
+	ds := newDictState(nil)
+
+	output.AttributeMatches = make([]mixerpb.ReferencedAttributes_AttributeMatch, len(pb.referencedAttrs))
+	i := 0
+	for k, v := range pb.referencedAttrs {
+		output.AttributeMatches[i] = mixerpb.ReferencedAttributes_AttributeMatch{
+			Name:      ds.getIndex(k),
```
Comment (ZackButcher) [inline pkg/attribute/protoBag.go]: Oh wow. It took me a while to understand what this is doing. I'm not a fan of `getIndex` mutating the underlying store, I think that kind of mutation might bite us down the road. In Java I'd call this method `getGlobalIndexOrCreateMessageIndex`. Maybe change this to `synthesizeIndex` or `getOrCreateIndex`, or similar, to denote that this is _not_ just a lookup in the dicts that the dictState was created with.

File: pkg/attribute/bag_test.go [hunk 2]
```diff
@@ -567,6 +568,119 @@ func TestMessageDictEdge(t *testing.T) {
...
+				var err error
+				b, err = GetBagFromProto(&attrs, globalWordList)
+				if err != nil {
+					t.Error("Expecting success, got %v", err)
+				}
+			}
+
+			v, ok := b.Get("HELLO")
+			if !ok {
+				t.Error("Expecting true, got false")
```
Comment (douglas-reid) [inline pkg/attribute/bag_test.go]: nit: error messages should be `got`, then `want`

File: pkg/attribute/bag_test.go [hunk 3]
```diff
@@ -567,6 +568,119 @@ func TestMessageDictEdge(t *testing.T) {
...
+
+func TestReferenceTracking(t *testing.T) {
+	globalWordList := []string{"G0", "G1", "G2"}
+	globalDict := map[string]int32{globalWordList[0]: 0, globalWordList[1]: 1, globalWordList[2]: 2}
+	messageWordList := []string{"N1", "N2", "N3"}
+
+	attrs := mixerpb.Attributes{
+		Words: messageWordList,
+		Strings: map[int32]int32{
+			0:  0,
```
Comment (douglas-reid) [inline pkg/attribute/bag_test.go]: can we add comments about the intended affect of these maps? I find it hard to follow.

File: pkg/attribute/dictState.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+		return index
+	}
+
+	if ds.messageDict == nil {
+		ds.messageDict = make(map[string]int32)
+	} else if index, ok := ds.messageDict[word]; ok {
+		return index
+	}
+
+	index := -int32(len(ds.messageDict)) - 1
```
Comment (douglas-reid) [inline pkg/attribute/dictState.go]: i feel like we need some helper method around index calculation. I think coming back to this later will leave us scratching our heads.

File: pkg/attribute/protoBag.go [hunk 5]
```diff
@@ -50,9 +64,56 @@ func (pb *ProtoBag) Get(name string) (interface{}, bool) {
...
+	return result, ok
+}
+
+// Get the set of attributes that have been referenced through this bag.
+func (pb *ProtoBag) GetReferencedAttributes() mixerpb.ReferencedAttributes {
+	output := mixerpb.ReferencedAttributes{}
+
+	ds := newDictState(nil)
+
+	output.AttributeMatches = make([]mixerpb.ReferencedAttributes_AttributeMatch, len(pb.referencedAttrs))
```
Comment (douglas-reid) [inline pkg/attribute/protoBag.go]: any reason not use use `make([]..., 0, len(pb.referencedAttrs))` and `append`s (instead of `i`) ?

File: pkg/attribute/protoBag.go [hunk 6]
```diff
@@ -33,14 +33,28 @@ type ProtoBag struct {
 	messageDict         map[string]int32
 	convertedStringMaps map[int32]map[string]string
 	stringMapMutex      sync.RWMutex
+
+	// to keep track of attributes that are referenced
+	referencedAttrs      map[string]mixerpb.ReferencedAttributes_Condition
```
Comment (douglas-reid) [inline pkg/attribute/protoBag.go]: candidate for `syncmap`? https://godoc.org/golang.org/x/sync/syncmap

File: pkg/attribute/bag_test.go [hunk 7]
```diff
@@ -567,6 +568,119 @@ func TestMessageDictEdge(t *testing.T) {
...
+	globalWordList := []string{"HELLO"}
+	globalDict := map[string]int32{globalWordList[0]: 0}
+	messageWordList := []string{"HELLO", "GOOD", "BAD"}
+
+	attrs := mixerpb.Attributes{
+		Words:   messageWordList,
+		Strings: map[int32]int32{0: -3, -1: -2},
+	}
+
+	for i := 0; i < 2; i++ {
```
Comment (geeknoid) [inline pkg/attribute/bag_test.go]: Cleaned up and commented, hopefully clearer.

File: pkg/attribute/bag_test.go [hunk 8]
```diff
@@ -567,6 +568,119 @@ func TestMessageDictEdge(t *testing.T) {
...
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			var b Bag
+
+			if i == 0 {
+				b = NewProtoBag(&attrs, globalDict, globalWordList)
+			} else {
+				var err error
+				b, err = GetBagFromProto(&attrs, globalWordList)
+				if err != nil {
+					t.Error("Expecting success, got %v", err)
```
Comment (geeknoid) [inline pkg/attribute/bag_test.go]: Done

File: pkg/attribute/dictState.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+func newDictState(globalDict map[string]int32) *dictState {
+	return &dictState{
+		globalDict: globalDict,
+	}
+}
+
+func (ds *dictState) getIndex(word string) int32 {
+	if index, ok := ds.globalDict[word]; ok {
```
Comment (geeknoid) [inline pkg/attribute/dictState.go]: Nope, this is going in the opposite direction (producing an Attributes message) and there we want to aggressively use the global dictionary when possible.

File: pkg/attribute/dictState.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	index := -int32(len(ds.messageDict)) - 1
+	ds.messageDict[word] = index
+	return index
+}
+
+func (ds *dictState) getMessageWordList() []string {
+	if len(ds.messageDict) == 0 {
+		return nil
```
Comment (geeknoid) [inline pkg/attribute/dictState.go]: I want nil here such that it matches a blank Attributes message.

File: pkg/attribute/protoBag.go (hunk 1 above)
Comment (geeknoid) [inline pkg/attribute/protoBag.go]: Renamed to assignIndex. Added big comment too.

File: pkg/attribute/bag_test.go (hunk 2 above)
Comment (geeknoid) [inline pkg/attribute/bag_test.go]: Done

File: pkg/attribute/bag_test.go (hunk 3 above)
Comment (geeknoid) [inline pkg/attribute/bag_test.go]: Done. Added several test comments.

File: pkg/attribute/dictState.go (hunk 4 above)
Comment (geeknoid) [inline pkg/attribute/dictState.go]: Done.

File: pkg/attribute/protoBag.go (hunk 6 above)
Comment (geeknoid) [inline pkg/attribute/protoBag.go]: Eventually. For now, syncmap doesn't support a Len method which is really a bummer.

**Comment by geeknoid** [comment]
PTAL

File: pkg/attribute/dictState.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+		globalDict:      globalDict,
+		globalWordCount: globalWordCount,
+	}
+}
+
+// assignDictIndex determines whether a word is in the allowed subset of the
+// global word list or whether it should be added to the message word list
+// being accumulated. It then returns the appropriate dictionary index given
+// the selected word list.
+func (ds *dictState) assignDictIndex(word string) int32 {
```
Comment (ZackButcher) [inline pkg/attribute/dictState.go]: Much clearer, thanks.
